File: src/client/java/fr/hardel/asset_editor/client/javafx/lib/action/EditorActionGateway.java
package fr.hardel.asset_editor.client.javafx.lib.action;

import fr.hardel.asset_editor.client.state.ClientPackInfo;
import fr.hardel.asset_editor.client.state.ClientSessionState;
import fr.hardel.asset_editor.client.state.ClientWorkspaceState;
import fr.hardel.asset_editor.client.state.PendingClientAction;
import fr.hardel.asset_editor.network.pack.PackWorkspaceRequestPayload;
import fr.hardel.asset_editor.network.workspace.WorkspaceElementSnapshot;
import fr.hardel.asset_editor.network.workspace.WorkspaceMutationRequestPayload;
import fr.hardel.asset_editor.network.workspace.WorkspaceSyncPayload;
import fr.hardel.asset_editor.store.ElementEntry;
import fr.hardel.asset_editor.workspace.action.EditorAction;
import fr.hardel.asset_editor.workspace.registry.RegistryWorkspaceBinding;
import fr.hardel.asset_editor.workspace.registry.RegistryWorkspaceBindings;
import net.fabricmc.fabric.api.client.networking.v1.ClientPlayNetworking;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import net.minecraft.client.Minecraft;
import net.minecraft.client.multiplayer.ClientPacketListener;
import net.minecraft.core.HolderLookup;
import net.minecraft.core.Registry;
import net.minecraft.resources.Identifier;
import net.minecraft.resources.ResourceKey;

import java.util.List;
import java.util.Objects;
import java.util.UUID;

public final class EditorActionGateway {

    private static final Logger LOGGER = LoggerFactory.getLogger(EditorActionGateway.class);

    private final ClientSessionState sessionState;
    private final ClientWorkspaceState workspaceState;

    public EditorActionGateway(ClientSessionState sessionState, ClientWorkspaceState workspaceState) {
        this.sessionState = sessionState;
        this.workspaceState = workspaceState;
    }

    public <T> EditorActionResult dispatch(ResourceKey<Registry<T>> registry, Identifier target, EditorAction action) {
        EditorActionResult check = validate(target);
        if (check != null)
            return check;

        ClientPackInfo pack = workspaceState.packSelectionState().selectedPack();
        if (pack == null)
            return EditorActionResult.packRequired();

        ElementEntry<T> entry = workspaceState.elementStore().get(registry, target);
        if (entry == null)
            return EditorActionResult.error("error:element_not_found");

        UUID actionId = UUID.randomUUID();
        workspaceState.trackPendingAction(actionId, new PendingClientAction<>(actionId, pack.packId(), registry, target, entry));

        projectOptimistic(registry, target, entry, action);

        ClientPlayNetworking.send(new WorkspaceMutationRequestPayload(
            actionId,
            pack.packId(),
            registry.identifier(),
            target,
            action));
        return EditorActionResult.applied();
    }

    private <T> void projectOptimistic(ResourceKey<Registry<T>> registry, Identifier target,
        ElementEntry<T> entry, EditorAction action) {
        RegistryWorkspaceBinding<T> binding = RegistryWorkspaceBindings.get(registry.identifier());
        if (binding == null || binding.interpreter() == null)
            return;

        HolderLookup.Provider registries = clientRegistries();
        if (registries == null)
            return;

        try {
            ElementEntry<T> projected = binding.interpreter().apply(entry, action, registries);
            if (projected != null && !Objects.equals(projected, entry))
                workspaceState.elementStore().put(registry, target, projected);
        } catch (Exception e) {
            LOGGER.warn("Optimistic projection failed for {}: {}", target, e.getMessage());
        }
    }

    public void requestPackWorkspace(ResourceKey<?> registry) {
        ClientPackInfo pack = workspaceState.packSelectionState().selectedPack();
        if (pack == null || pack.packId().isBlank())
            return;
        ClientPlayNetworking.send(new PackWorkspaceRequestPayload(pack.packId(), registry.identifier()));
    }

    public void handleWorkspaceSync(WorkspaceSyncPayload payload) {
        if (!payload.mutationResponse()) {
            if (payload.snapshot() != null)
                applySnapshotIfCurrentPack(payload.packId(), payload.snapshot());
            return;
        }

        PendingClientAction<?> pending = workspaceState.removePendingAction(payload.actionId());
        if (pending == null)
            return;
        if (payload.accepted()) {
            if (payload.snapshot() != null)
                applySnapshotIfCurrentPack(payload.packId(), payload.snapshot());
            return;
        }
        restorePending(pending);
        workspaceState.issueState().pushError(payload.errorCode());
    }

    public void handlePackWorkspaceSync(String packId, Identifier registryId, List<WorkspaceElementSnapshot> snapshots) {
        ClientPackInfo selectedPack = workspaceState.packSelectionState().selectedPack();
        if (selectedPack == null || !selectedPack.packId().equals(packId))
            return;

        var binding = RegistryWorkspaceBindings.get(registryId);
        if (binding == null)
            return;

        HolderLookup.Provider registries = clientRegistries();
        if (registries == null)
            return;

        replaceAll(binding, snapshots, registries);
    }

    private void restorePending(PendingClientAction<?> pending) {
        ClientPackInfo selectedPack = workspaceState.packSelectionState().selectedPack();
        if (selectedPack == null || !selectedPack.packId().equals(pending.packId()))
            return;
        restorePendingTyped(pending);
    }

    private <T> void restorePendingTyped(PendingClientAction<T> pending) {
        workspaceState.elementStore().put(
            pending.registry(),
            pending.target(),
            pending.previousSnapshot());
    }

    private void applySnapshotIfCurrentPack(String packId, WorkspaceElementSnapshot snapshot) {
        ClientPackInfo selectedPack = workspaceState.packSelectionState().selectedPack();
        if (selectedPack == null || !selectedPack.packId().equals(packId))
            return;

        var binding = RegistryWorkspaceBindings.get(snapshot.registryId());
        if (binding == null)
            return;

        HolderLookup.Provider registries = clientRegistries();
        if (registries == null)
            return;

        applySnapshot(binding, snapshot, registries);
    }

    private <T> void replaceAll(RegistryWorkspaceBinding<T> binding, List<WorkspaceElementSnapshot> snapshots,
        HolderLookup.Provider registries) {
        List<ElementEntry<T>> entries = snapshots.stream()
            .map(snapshot -> binding.fromSnapshot(snapshot, registries))
            .toList();
        workspaceState.elementStore().replaceAll(binding.registryKey(), entries);
    }

    private <T> void applySnapshot(RegistryWorkspaceBinding<T> binding, WorkspaceElementSnapshot snapshot,
        HolderLookup.Provider registries) {
        ElementEntry<T> entry = binding.fromSnapshot(snapshot, registries);
        workspaceState.elementStore().put(binding.registryKey(), snapshot.targetId(), entry);
    }

    private HolderLookup.Provider clientRegistries() {
        ClientPacketListener connection = Minecraft.getInstance().getConnection();
        return connection != null ? connection.registryAccess() : null;
    }

    private EditorActionResult validate(Identifier target) {
        if (target == null)
            return EditorActionResult.error("error:element_not_found");
        var pack = workspaceState.packSelectionState().selectedPack();
        if (pack == null)
            return EditorActionResult.packRequired();
        if (!pack.writable())
            return EditorActionResult.rejected("error:pack_readonly");
        if (!sessionState.permissions().canEdit())
            return EditorActionResult.rejected("error:permission_denied");
        return null;
    }
}

File: src/client/kotlin/fr/hardel/asset_editor/client/compose/lib/action/EditorActionGateway.kt
package fr.hardel.asset_editor.client.compose.lib.action

import fr.hardel.asset_editor.client.WorkspaceSyncGateway
import fr.hardel.asset_editor.client.debug.ClientDebugTelemetry
import fr.hardel.asset_editor.client.memory.session.SessionMemory
import fr.hardel.asset_editor.client.memory.persistent.IssueMemory
import fr.hardel.asset_editor.client.memory.session.ui.ClientPackInfo
import fr.hardel.asset_editor.client.memory.session.server.ClientWorkspaceRegistries
import fr.hardel.asset_editor.client.memory.session.server.ClientWorkspaceRegistry
import fr.hardel.asset_editor.client.memory.session.ui.PackSelectionMemory
import fr.hardel.asset_editor.client.memory.session.server.RegistryMemory
import fr.hardel.asset_editor.client.network.ClientPayloadSender
import fr.hardel.asset_editor.network.pack.PackWorkspaceRequestPayload
import fr.hardel.asset_editor.network.workspace.ElementSeedRequestPayload
import fr.hardel.asset_editor.network.workspace.WorkspaceElementSnapshot
import fr.hardel.asset_editor.network.workspace.WorkspaceMutationRequestPayload
import fr.hardel.asset_editor.network.workspace.WorkspaceSyncPayload
import fr.hardel.asset_editor.workspace.flush.ElementEntry
import fr.hardel.asset_editor.workspace.action.EditorAction

import fr.hardel.asset_editor.workspace.io.RegistryMutationContexts
import java.util.Objects
import java.util.UUID
import net.minecraft.client.Minecraft
import net.minecraft.client.multiplayer.ClientPacketListener
import net.minecraft.core.HolderLookup
import net.minecraft.resources.Identifier
import org.slf4j.LoggerFactory

class EditorActionGateway(
    private val sessionMemory: SessionMemory,
    private val packSelection: PackSelectionMemory,
    private val registryMemory: RegistryMemory,
    private val issues: IssueMemory
) : WorkspaceSyncGateway {

    private val pendingActions = LinkedHashMap<UUID, PendingClientAction<*>>()

    fun pendingActionCount(): Int = pendingActions.size

    fun clearPendingActions() {
        pendingActions.clear()
    }

    private fun trackPendingAction(actionId: UUID, action: PendingClientAction<*>) {
        pendingActions[actionId] = action
    }

    private fun removePendingAction(actionId: UUID): PendingClientAction<*>? {
        return pendingActions.remove(actionId)
    }

    fun <T : Any> dispatch(
        workspace: ClientWorkspaceRegistry<T>,
        target: Identifier?,
        action: EditorAction<*>
    ): EditorActionResult {
        val check = validate(target)
        if (check != null) {
            return check
        }

        val pack = packSelection.selectedPack()
            ?: return EditorActionResult.packRequired()
        val resolvedTarget = target ?: return EditorActionResult.error("error:workspace_sync_pending")
        val entry = registryMemory.get(workspace, resolvedTarget)
            ?: return EditorActionResult.error("error:workspace_sync_pending")

        val actionId = UUID.randomUUID()
        trackPendingAction(
            actionId,
            PendingClientAction(actionId, pack.packId(), workspace, resolvedTarget, entry)
        )

        projectOptimistic(workspace, resolvedTarget, entry, action)
        val payload = WorkspaceMutationRequestPayload(
            actionId,
            pack.packId(),
            workspace.registryId(),
            resolvedTarget,
            action
        )
        ClientDebugTelemetry.actionDispatched(pack.packId(), workspace.registryId(), resolvedTarget, action, actionId)
        ClientPayloadSender.send(payload)
        return EditorActionResult.applied()
    }

    fun requestPackWorkspace(workspace: ClientWorkspaceRegistry<*>) {
        val pack = packSelection.selectedPack()
        if (pack == null || pack.packId().isBlank()) return
        ClientPayloadSender.send(PackWorkspaceRequestPayload(pack.packId(), workspace.registryId()))
    }

    fun requestElementSeed(workspace: ClientWorkspaceRegistry<*>, elementId: Identifier) {
        val pack = packSelection.selectedPack()
        if (pack == null || pack.packId().isBlank()) return
        ClientPayloadSender.send(ElementSeedRequestPayload(pack.packId(), workspace.registryId(), elementId))
    }

    override fun handleWorkspaceSync(payload: WorkspaceSyncPayload) {
        if (!payload.mutationResponse()) {
            payload.snapshot()?.let { applySnapshotIfCurrentPack(payload.packId(), it) }
            return
        }

        val pending = removePendingAction(payload.actionId()) ?: return
        if (payload.accepted()) {
            payload.snapshot()?.let { applySnapshotIfCurrentPack(payload.packId(), it) }
            return
        }

        restorePending(pending)
        ClientDebugTelemetry.actionRejected(payload.actionId(), payload.errorCode() ?: "unknown")
        issues.pushError(payload.errorCode())
    }

    override fun handlePackWorkspaceSync(
        packId: String,
        registryId: Identifier,
        snapshots: List<WorkspaceElementSnapshot>
    ) {
        val selectedPack = packSelection.selectedPack()
        if (selectedPack == null || selectedPack.packId() != packId) {
            return
        }

        val workspace = ClientWorkspaceRegistries.get(registryId) ?: return
        val registries = clientRegistries() ?: return
        replaceAll(workspace, snapshots, registries)
    }

    private fun <T : Any> projectOptimistic(
        workspace: ClientWorkspaceRegistry<T>,
        target: Identifier,
        entry: ElementEntry<T>,
        action: EditorAction<*>
    ) {
        val registries = clientRegistries() ?: return

        try {
            val context = RegistryMutationContexts.client(registries)
            val projected = workspace.definition().apply(entry, action, context)
            if (!Objects.equals(projected, entry)) {
                registryMemory.put(workspace, target, projected)
            }
        } catch (exception: Exception) {
            LOGGER.warn("Optimistic projection failed for {}: {}", target, exception.message)
            ClientDebugTelemetry.optimisticFailed(workspace.registryId(), target, action, exception.message ?: "unknown")
        }
    }

    private fun restorePending(pending: PendingClientAction<*>) {
        val selectedPack = packSelection.selectedPack()
        if (selectedPack == null || selectedPack.packId() != pending.packId) {
            return
        }

        pending.restoreInto(registryMemory)
    }

    private fun applySnapshotIfCurrentPack(packId: String, snapshot: WorkspaceElementSnapshot) {
        val selectedPack = packSelection.selectedPack()
        if (selectedPack == null || selectedPack.packId() != packId) {
            return
        }

        val workspace = ClientWorkspaceRegistries.get(snapshot.registryId()) ?: return
        val registries = clientRegistries() ?: return
        applySnapshot(workspace, snapshot, registries)
    }

    private fun replaceAll(
        workspace: ClientWorkspaceRegistry<*>,
        snapshots: List<WorkspaceElementSnapshot>,
        registries: HolderLookup.Provider
    ) {
        replaceAllTyped(workspace, snapshots, registries)
    }

    private fun <T : Any> replaceAllTyped(
        workspace: ClientWorkspaceRegistry<T>,
        snapshots: List<WorkspaceElementSnapshot>,
        registries: HolderLookup.Provider
    ) {
        val entries = snapshots.map { workspace.definition().fromSnapshot(it, registries) }
        registryMemory.replaceAll(workspace, entries)
    }

    private fun applySnapshot(
        workspace: ClientWorkspaceRegistry<*>,
        snapshot: WorkspaceElementSnapshot,
        registries: HolderLookup.Provider
    ) {
        applySnapshotTyped(workspace, snapshot, registries)
    }

    private fun <T : Any> applySnapshotTyped(
        workspace: ClientWorkspaceRegistry<T>,
        snapshot: WorkspaceElementSnapshot,
        registries: HolderLookup.Provider
    ) {
        val entry = workspace.definition().fromSnapshot(snapshot, registries)
        registryMemory.put(workspace, snapshot.targetId(), entry)
    }

    private fun clientRegistries(): HolderLookup.Provider? {
        val connection: ClientPacketListener? = Minecraft.getInstance().connection
        return connection?.registryAccess()
    }

    private fun validate(target: Identifier?): EditorActionResult? {
        if (target == null) {
            return EditorActionResult.error("error:workspace_sync_pending")
        }

        val pack: ClientPackInfo = packSelection.selectedPack()
            ?: return EditorActionResult.packRequired()
        if (!pack.writable()) {
            return EditorActionResult.rejected("error:pack_readonly")
        }
        if (!sessionMemory.permissions().canEdit()) {
            return EditorActionResult.rejected("error:permission_denied")
        }

        return null
    }

    data class PendingClientAction<T : Any>(
        val actionId: UUID,
        val packId: String,
        val workspace: ClientWorkspaceRegistry<T>,
        val target: Identifier,
        val previousSnapshot: ElementEntry<T>
    ) {
        fun restoreInto(memory: RegistryMemory) {
            memory.put(workspace, target, previousSnapshot)
        }
    }

    companion object {
        private val LOGGER = LoggerFactory.getLogger(EditorActionGateway::class.java)
    }
}
